dan_layer/transaction/src/id_provider.rs

```diff
@@ -71,8 +71,14 @@ impl IdProvider {
             .into())
     }
 
-    pub fn new_component_address(&self) -> Result<ComponentAddress, IdProviderError> {
-        Ok(self.new_id()?.into())
+    pub fn new_component_address(
+        &self,
+        template_address: TemplateAddress,
+        component_id: Option<Hash>,
+    ) -> Result<ComponentAddress, IdProviderError> {
+        // if the component_id is not specified by the caller, then it will be random
+        let component_id = component_id.unwrap_or(self.new_id()?);
```

Comment:
Surprised clippy didn't catch this https://rust-lang.github.io/rust-clippy/v0.0.212/index.html#or_fun_call
This will increment the counter in every case.

Reply:
Nice catch, I forgot that `unwrap_or` is eagerly evaluated